src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -149,15 +176,355 @@
+        voucher.setAmount(calculatedAmount);
+        return handleVoucher(voucher)
+          .thenAccept(voucherWithId -> populateVoucherId(voucherLines, voucher))
+          .thenCompose(v -> createVoucherLinesRecords(voucherLines));
+      });
+  }
+
+  /**
+   * Prepares the data necessary for the generation of voucher lines based on the invoice lines found
+   *
+   * @param invoiceLines {@link List<InvoiceLine>} associated with processed {@link Invoice}
+   * @return {@link InvoiceLine#fundDistributions} grouped by {@link Fund#externalAccountNo}
+   */
+  private CompletableFuture<Map<String, List<FundDistribution>>> groupFundDistributionsByExternalAccountNo(
+      List<InvoiceLine> invoiceLines) {
+
+    Map<String, List<FundDistribution>> fundDistributionsGroupedByFundId = getGroupFundDistributionsByFundId(invoiceLines);
+
+    return fetchFoundsByIds(new ArrayList<>(fundDistributionsGroupedByFundId.keySet()))
+      .thenApply(this::groupFundsByExternalAccountNumber)
+      .thenApply(fundsGroupedByExternalAccountNo ->
+        mapExternalAccountNumberToFundDistributions(fundDistributionsGroupedByFundId, fundsGroupedByExternalAccountNo)
+      );
+  }
+
+  private Map<String, List<Fund>> groupFundsByExternalAccountNumber(List<Fund> funds) {
+    return funds.stream().collect(groupingBy(Fund::getExternalAccountNo));
+  }
+
+  private Map<String, List<FundDistribution>> mapExternalAccountNumberToFundDistributions(
+    Map<String, List<FundDistribution>> fundDistributionsGroupedByFundId,
+    Map<String, List<Fund>> fundsGroupedByExternalAccountNo) {
+
+    return fundsGroupedByExternalAccountNo.keySet()
+      .stream()
+      .collect(toMap(externalAccountNo -> externalAccountNo,
+        externalAccountNo -> fundsGroupedByExternalAccountNo.get(externalAccountNo)
+          .stream()
+          .map(Fund::getId)
+          .flatMap(fundId -> fundDistributionsGroupedByFundId.get(fundId)
+            .stream())
+          .collect(toList())));
+  }
+
+  private Map<String, List<FundDistribution>> getGroupFundDistributionsByFundId(List<InvoiceLine> invoiceLines) {
+    return invoiceLines.stream()
+      .flatMap(invoiceLine -> invoiceLine.getFundDistributions().stream()
+        .map(fundDistribution -> fundDistribution.withInvoiceLineId(invoiceLine.getId())))
+      .collect(groupingBy(FundDistribution::getFundId));
+  }
+
+  private CompletableFuture<List<Fund>> fetchFoundsByIds(List<String> fundIds) {
+    List<CompletableFuture<List<Fund>>> futures = StreamEx
+      .ofSubLists(fundIds, MAX_IDS_FOR_GET_RQ)
+      // Send get request for each CQL query
+      .map(this::getFoundsByIds)
+      .collect(toList());
+
+    return collectResultsOnSuccess(futures)
+      .thenApply(results -> results
+        .stream()
+        .flatMap(List::stream)
+        .collect(toList())
+      )
+      .thenApply(existedFunds -> verifyThatAllFundsFound(existedFunds, fundIds));
```

Comment:
nitpick...  using the present tense makes more sense IMO `existingFunds` instead of `existedFunds`

Reply:
corrected